R/1.0_module_data_summary.R

```diff
@@ -0,0 +1,255 @@
+        attr(filter_overview$dataname, "label") <- "Data Name"
+
+        if (!is.null(filter_overview$obs)) {
+          # some datasets (MAE colData) doesn't return obs column
+          filter_overview <- transform(
+            filter_overview,
+            obs_str_summary = ifelse(
+              !is.na(obs),
+              sprintf("%s/%s", obs_filtered, obs),
+              ""
+            )
+          )
+          attr(filter_overview$obs_str_summary, "label") <- "Obs"
+        }
+
+
+        if (!is.null(filter_overview$subjects)) {
+          # some datasets (without keys) doesn't return subjects
+          filter_overview <- transform(
+            filter_overview,
+            subjects_summary = ifelse(
+              !is.na(subjects),
+              sprintf("%s/%s", subjects_filtered, subjects),
+              ""
+            )
+          )
+          attr(filter_overview$subjects_summary, "label") <- "Subjects"
+        }
+
+        all_names <- c("dataname", "obs_str_summary", "subjects_summary")
+        filter_overview <- filter_overview[, colnames(filter_overview) %in% all_names]
+
+        body_html <- apply(
+          filter_overview,
+          1,
+          function(x) {
+            tags$tr(
+              tagList(
+                tags$td(
+                  if (all(x[-1] == "")) {
+                    icon(
+                      name = "exclamation-triangle",
+                      title = "Unsupported dataset",
+                      `data-container` = "body",
+                      `data-toggle` = "popover",
+                      `data-content` = "object not supported by the data_summary module"
+                    )
+                  },
+                  x[1]
+                ),
+                lapply(x[-1], tags$td)
+              )
+            )
+          }
+        )
+
+        header_labels <- vapply(
+          seq_along(filter_overview),
+          function(i) {
+            label <- attr(filter_overview[[i]], "label")
+            ifelse(!is.null(label), label, names(filter_overview)[[i]])
+          },
+          character(1)
+        )
+        header_html <- tags$tr(tagList(lapply(header_labels, tags$td)))
+
+        table_html <- tags$table(
+          class = "table custom-table",
+          tags$thead(header_html),
+          tags$tbody(body_html)
+        )
+        logger::log_trace("srv_data_summary updated counts")
+        table_html
+      })
+      logger::log_trace("srv_data_summary initialized")
+      NULL
+    }
+  )
+}
+
+get_filter_overview <- function(filtered_teal_data) {
+  rows <- lapply(
+    datanames(filtered_teal_data()),
+    get_object_filter_overview,
+    filtered_teal_data = filtered_teal_data
+  )
+  unssuported_idx <- vapply(rows, function(x) all(is.na(x[-1])), logical(1)) # this is mainly for vectors
+  dplyr::bind_rows(c(rows[!unssuported_idx], rows[unssuported_idx]))
+}
+
+get_object_filter_overview <- function(filtered_teal_data, dataname, experiment_name = NULL) {
```

Comment:
`filtered_teal_data` being a `teal_data` complicates this dispatch. I recommend following design:

```r
get_filter_overview <- function(teal_data) {
  datanames <- teal.data::datanames(teal_data)
  filtered_data_objs <- <get filtered objects>
  unfiltered_data_objs <- <get unfiltered objects>

  lapply(
    datanames,
    function(dataname) {
      get_object_filter_overview(
        filtered_data = filtered_data_objs[[dataname]],
        unfiltered_data = unfiltered_data_objs[[dataname]]
    }
  )
}

get_object_filter_overview <- function(filtered_data, unfiltered_data, dataname) {
  
}
```

Benefit of such `get_object_filter_overview` is that it will be called recursively and we will be able to register other "pseudo S3" for any class.

Reply:
Thanks for the help! Will try to improve